clippy_lints/src/if_then_some_else_none.rs

```diff
@@ -0,0 +1,115 @@
+use crate::utils;
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_semver::RustcVersion;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+
+const IF_THEN_SOME_ELSE_NONE_MSRV: RustcVersion = RustcVersion::new(1, 50, 0);
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for if-else that could be written to `bool::then`.
+    ///
+    /// **Why is this bad?** Looks a little redundant. Using `bool::then` helps it have less lines of code.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// # let v = vec![0];
+    /// let a = if v.is_empty() {
+    ///     println!("true!");
+    ///     Some(42)
+    /// } else {
+    ///     None
+    /// };
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```rust
+    /// # let v = vec![0];
+    /// let a = v.is_empty().then(|| {
+    ///     println!("true!");
+    ///     42
+    /// });
+    /// ```
+    pub IF_THEN_SOME_ELSE_NONE,
```

Comment:
Since the MSRV feature is pretty new an we don't have any automated checks for it yet:

If an MSRV is added to a lint, the lint has also to be added in the config option. I wrote documentation about this in #6977. cc @rust-lang/clippy

Reply:
I will prepare a fix for this (and possible other lints soon.

Reply:
> If an MSRV is added to a lint, the lint has also to be added in the config option.

Oh I didn't know that. I'll keep that in mind for the next time I implement a new lint.

> I will prepare a fix for this (and possible other lints soon.

Thanks!

Reply:
Thanks for your follow-up!

Reply:
> Oh I didn't know that. I'll keep that in mind for the next time I implement a new lint.

No worries, this isn't/wasn't documented anywhere, so that's on me for missing it while reviewing the documentation.